#### Life cycle of Machine learning Project

- Understanding the Problem Statement
- Data Collection
- Data Checks to perform
- Exploratory data analysis
- Data Pre-Processing
- Model Training
- Choose best model

### Students Performance in Exams Dataset

The **Students Performance in Exams** dataset contains information about students and their exam scores.
It includes students' **gender, parental education, lunch type, and test preparation course**.
The dataset also contains their scores in **math, reading, and writing**.
It can be used to study **what factors may affect students' academic performance**.

| Feature Name                  | Meaning                                            | Easy Explanation                                                                                                               |
| ----------------------------- | -------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------ |
| `gender`                      | Gender of the student                              | Tells whether the student is **male or female**.                                                                               |
| `race/ethnicity`              | Race or ethnicity group                            | Shows which **group/category** the student belongs to, such as Group A, B, C, D, or E.                                         |
| `parental level of education` | Highest education level of the student's parents   | Shows how much **education the student's parents have completed**, such as high school, bachelor's degree, or master's degree. |
| `lunch`                       | Type of lunch received by the student              | Shows whether the student receives **standard lunch** or **free/reduced lunch**.                                               |
| `test preparation course`     | Whether the student completed a preparation course | Shows whether the student **completed a test preparation course** or did not take one.                                         |
| `math score`                  | Student's mathematics exam score                   | Shows how many **marks the student obtained in the math exam**.                                                                |
| `reading score`               | Student's reading exam score                       | Shows how many **marks the student obtained in the reading exam**.                                                             |
| `writing score`               | Student's writing exam score                       | Shows how many **marks the student obtained in the writing exam**.                                                             |


### 1) Problem statement
- This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.


### 2) Data Collection
- Dataset Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
- The data consists of 8 column and 1000 rows.

### 2.1 Import Data and Required Packages
####  Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [42]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

#### Import the CSV Data as Pandas DataFrame

In [43]:
df = pd.read_csv('data/StudentsPerformance.csv')

#### Show Top 5 Records

In [44]:
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


#### Shape of the dataset

In [45]:
df.shape

(1000, 8)

### 2.2 Dataset Information.

- gender : sex of students  -> (Male/female)
- race/ethnicity : ethnicity of students -> (Group A, B,C, D,E)
- parental level of education : parents' final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)
- lunch : having lunch before test (standard or free/reduced) 
- test preparation course : complete or not complete before test
- math score
- reading score
- writing score

### 3. Data Checks to perform

- Check Missing values
- Check Duplicates
- Check data type
- Check the number of unique values of each column
- Check statistics of data set
- Check various categories present in the different categorical column

### 3.1 Check Missing Values

In [46]:
df.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

### Insights
No Null Values in the dataset

### 3.2 Check Duplicates

In [47]:
# df[df.duplicated()]    
df.duplicated().sum()

0

### insights
No duplicate Values in the dataset


### 3.3 Checking Data Type

In [48]:
df.dtypes

gender                         object
race/ethnicity                 object
parental level of education    object
lunch                          object
test preparation course        object
math score                      int64
reading score                   int64
writing score                   int64
dtype: object

### 3.4 Checking Unique Values in each Column.

In [49]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

### 3.5 Statistical Summary of the Dataset.

In [50]:
df.describe()

,math score,reading score,writing score
count,1000.00000,1000.000000,1000.000000
mean,66.08900,69.169000,68.054000
std,15.16308,14.600192,15.195657
min,0.00000,17.000000,10.000000
25%,57.00000,59.000000,57.750000
50%,66.00000,70.000000,69.000000
75%,77.00000,79.000000,79.000000
max,100.00000,100.000000,100.000000


### Insights
- All mean values are very close to each other.
- All std values are very close to each other range 14.6 to 15.1.
- Minimum score for maths,reading,writing are 0,10,17.

In [51]:
catagorical_columns = [col for col in df.columns if df[col].dtypes == 'O']
print(len(catagorical_columns))

5


### 3.6 mCheck various categories present in the different categorical column

In [52]:
print("Categories in 'Gender' Column:",end=' ')
print(df['gender'].unique())
print('\n')
print("Categories in 'race/ethnicity' Column:",end=' ')
print(df['race/ethnicity'].unique())
print('\n')
print("Categories in'parental level of education' Column:",end=" " )
print(df['parental level of education'].unique())
print('\n')
print("Categories in 'lunch' Column:     ",end=" " )
print(df['lunch'].unique())
print('\n')
print("Categories in 'test preparation course' Column:     ",end=" " )
print(df['test preparation course'].unique())

Categories in 'Gender' Column: ['female' 'male']


Categories in 'race/ethnicity' Column: ['group B' 'group C' 'group A' 'group D' 'group E']


Categories in'parental level of education' Column: ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']


Categories in 'lunch' Column:      ['standard' 'free/reduced']


Categories in 'test preparation course' Column:      ['none' 'completed']


### 3.7 Check the Number of Numerical and Categorical Features.

In [61]:
categorical = [col for col in df.columns if df[col].dtypes == 'O']
numerical = [col for col in df.columns if df[col].dtypes != 'O']

print(f"We have {len(categorical)} Categorical Columns in this dataframe \n {categorical}")
print("\n")
print(f"We have {len(numerical)} Numerical Columns in this dataframe \n {numerical}")

We have 5 Categorical Columns in this dataframe 
 ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']


We have 3 Numerical Columns in this dataframe 
 ['math score', 'reading score', 'writing score']


In [72]:
df = df.rename(columns={'parental level of education':'parental_level_of_education','test preparation course':'test_preparation_course'})

### 3.8 Adding columns for "Total Score" and "Average"

In [73]:
df['Total Score'] = df['math score'] + df['reading score'] + df['writing score']
df['Average'] = df['Total Score']/3

In [74]:
df.head()

,gender,race/ethnicity,parental_level_of_education,lunch,test_preparation_course,math score,reading score,writing score,Total Score,Average
0,female,group B,bachelor's degree,standard,none,72,72,74,218,72.666667
1,female,group C,some college,standard,completed,69,90,88,247,82.333333
2,female,group B,master's degree,standard,none,90,95,93,278,92.666667
3,male,group A,associate's degree,free/reduced,none,47,57,44,148,49.333333
4,male,group C,some college,standard,none,76,78,75,229,76.333333


In [85]:
reading_full =df[df['reading score'] == 100]['Average'].count()
writing_full =df[df['writing score'] == 100]['Average'].count()
math_full =df[df['math score'] == 100]['Average'].count()

print('Number of Students with full marks in Maths',math_full)
print('Number of Students with full marks in Writing',writing_full)
print('Number of Students with full marks in Reading',reading_full)

Number of Students with full marks in Maths 7
Number of Students with full marks in Writing 14
Number of Students with full marks in Reading 17


In [87]:
reading_less =df[df['reading score'] < 20]['Average'].count()
writing_less =df[df['writing score'] < 20]['Average'].count()
math_less =df[df['math score'] < 20]['Average'].count()

print('Number of Students with less than 20 marks in Maths',math_less)
print('Number of Students with less than 20 marks in Writing',writing_less)
print('Number of Students with less than 20 marks in  Reading',reading_less)

Number of Students with less than 20 marks in Maths 4
Number of Students with less than 20 marks in Writing 3
Number of Students with less than 20 marks in  Reading 1


#####  Insights
 - From above values we get students have performed the worst in Maths 
 - Best performance is in reading section